tariff/octopus/graphql/types.go

```diff
@@ -0,0 +1,78 @@
+package graphql
+
+// BaseURI is Octopus Energy's core API root.
+const BaseURI = "https://api.octopus.energy"
+
+// URI is the GraphQL query endpoint for Octopus Energy.
+const URI = BaseURI + "/v1/graphql/"
+
+// FIXME these don't need to be public
+
+// KrakenTokenAuthentication is a representation of a GraphQL query for obtaining a Kraken API token.
+type KrakenTokenAuthentication struct {
+	ObtainKrakenToken struct {
+		Token string
+	} `graphql:"obtainKrakenToken(input: {APIKey: $apiKey})"`
+}
+
+// KrakenAccountLookup is a representation of a GraphQL query for obtaining the Account Number associated with the
+// credentials used to authorize the request.
+type KrakenAccountLookup struct {
+	Viewer struct {
+		Accounts []struct {
+			Number string
+		}
+	}
+}
+
+type tariffType struct {
+	Id                   string
+	DisplayName          string
+	FullName             string
+	ProductCode          string
+	StandingCharge       float32
+	PreVatStandingCharge float32
+}
+
+type tariffTypeWithTariffCode struct {
+	tariffType
+	TariffCode string
+}
+
+type StandardTariff struct {
+	tariffTypeWithTariffCode
+}
+type DayNightTariff struct {
+	tariffTypeWithTariffCode
+}
+type ThreeRateTariff struct {
+	tariffTypeWithTariffCode
+}
+type HalfHourlyTariff struct {
+	tariffTypeWithTariffCode
+}
+type PrepayTariff struct {
+	tariffTypeWithTariffCode
+}
+
+type KrakenAccountElectricityAgreements struct {
+	Account struct {
+		ElectricityAgreements []struct {
+			Id     int
+			Tariff struct {
+				// yukky but the best way I can think of to handle this
+				// access via any relevant tariff data entry (i.e. StandardTariff)
+				// TODO would appreciate peer review
+				StandardTariff   `graphql:"... on StandardTariff"`
+				DayNightTariff   `graphql:"... on DayNightTariff"`
+				ThreeRateTariff  `graphql:"... on ThreeRateTariff"`
+				HalfHourlyTariff `graphql:"... on HalfHourlyTariff"`
+				PrepayTariff     `graphql:"... on PrepayTariff"`
```

Comment:
Would appreciate your thoughts on whether this is the best way to handle this.

Different types are returned, but the necessary `tariffCode` is only available inside the derivative types, not inside the generic `tariffType`. See API docs @ https://api.octopus.energy/v1/graphql/

Reply:
GraphQL puzzles me every time I look at it. No idea :/

Reply:
The only other option I can see would be to use a [custom query](https://github.com/hasura/go-graphql-client#execute-pre-built-query) and build the query string from scratch, but that scratches me as an improper way to do it - and potentially not type safe. Leaning towards going forward with the present method with data retrieval functions to take the yukiness out of it.